File: app-figlio/app/src/main/java/eu/stgm/pactum/figlio/valutatore/Valutatore.kt
package eu.stgm.pactum.figlio.valutatore

import eu.stgm.pactum.figlio.dati.Regola
import eu.stgm.pactum.figlio.dati.TipiRegola
import eu.stgm.pactum.figlio.misura.UsoApp
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.time.Instant
import java.time.LocalTime
import java.time.ZoneId
import java.time.format.DateTimeParseException

/**
 * Uno sforamento rilevato sul telefono (contratto-api.md, evento `sforamento`).
 * `limiteEfficace` è valorizzato solo per limite_tempo (= minuti_al_giorno +
 * bonus di oggi su quella regola); per le fasce orarie non c'è un tetto di
 * minuti, `minutiOltre` racconta quanti minuti d'uso sono caduti nella fascia.
 * `giornoAncora` è valorizzato solo per le fasce: è il giorno di ANCORAGGIO
 * dell'occorrenza (quando la fascia parte) e fa da chiave di dedup, così una
 * fascia che scavalca la mezzanotte resta UNA sola occorrenza.
 */
data class Sforamento(
    val regolaId: Long,
    val tipo: String,
    val limiteEfficace: Int?,
    val minutiOltre: Int,
    val giornoAncora: String? = null,
)

/** Un intervallo proibito da una fascia, con il giorno di ancoraggio dell'occorrenza. */
data class IntervalloProibito(val giornoAncora: String, val inizio: Long, val fine: Long)

/** Dove sta oggi una fascia oraria rispetto ad adesso. */
sealed interface MomentoFascia {
    /** La fascia parte più tardi, oggi: mancano [minuti] alle [inizio]. */
    data class Prima(val minuti: Long, val inizio: LocalTime) : MomentoFascia

    /** Adesso sei dentro la fascia, fino alle [fine]. */
    data class InCorso(val minuti: Long, val fine: LocalTime) : MomentoFascia

    /** Oggi la fascia c'è stata ed è finita. */
    data object Finita : MomentoFascia

    /** Oggi la fascia non vale. */
    data object NonOggi : MomentoFascia
}

/**
 * L'uso di oggi indicizzato come lo legge il valutatore: per pacchetto (in
 * minuscolo) e per chiave `categoria:*`, con i minuti arrotondati per app —
 * lo stesso conto di SentinellaPatto e della schermata Oggi, così "48 min su
 * 1 h" e lo sforamento parlano degli stessi minuti.
 */
class IndiceUso(
    uso: List<Pair<String, Long>>,
    categoriaDi: (String) -> String,
) {
    private val perPacchetto = HashMap<String, Long>()
    private val perCategoria = HashMap<String, Long>()

    init {
        for ((pacchetto, millis) in uso) {
            val minuti = millis / 60_000
            perPacchetto.merge(pacchetto.lowercase(), minuti, Long::plus)
            perCategoria.merge(categoriaDi(pacchetto), minuti, Long::plus)
        }
    }

    /** I minuti di oggi su una chiave `app_o_categoria` (match esatto, contratto v2.1). */
    fun minuti(chiave: String): Long {
        val k = chiave.trim().lowercase()
        return if (k.startsWith(PREFISSO_CATEGORIA)) perCategoria[k] ?: 0L else perPacchetto[k] ?: 0L
    }

    companion object {
        private const val PREFISSO_CATEGORIA = "categoria:"

        /**
         * L'indice dall'uso letto sul telefono, tenendo SOLO i pacchetti che
         * [contaNellUso] ammette: lo stesso filtro della fotografia inviata al
         * server, così una categoria qui somma le stesse app che il genitore
         * vede sommate nella sua finestra.
         */
        fun daUso(
            uso: List<UsoApp>,
            contaNellUso: (String) -> Boolean,
            categoriaDi: (String) -> String,
        ): IndiceUso = IndiceUso(
            uso = uso.filter { contaNellUso(it.pacchetto) }.map { it.pacchetto to it.millisPrimoPiano },
            categoriaDi = categoriaDi,
        )
    }
}

/**
 * Il valutatore locale: confronta l'uso di oggi con le regole attive del patto e
 * restituisce gli sforamenti. È logica pura (nessun IO): l'uso arriva da due
 * lambde — per chiave app_o_categoria (limite_tempo) e per intervallo di tempo
 * (fascia_oraria) — così è testabile e riusabile dal worker e dal loop del
 * servizio. NON blocca niente e NON deduplica: la deduplica (una per regola per
 * giorno, o per giorno di ancoraggio per le fasce) e la notifica le fa il chiamante.
 */
object Valutatore {

    // Ordine ISO: lun=1 … dom=7 (LocalDate.dayOfWeek.value), indicizzato da 0.
    private val GIORNI = listOf("lun", "mar", "mer", "gio", "ven", "sab", "dom")

    /** Sotto il minuto d'uso in una fascia = rumore di misura, non uno sforamento. */
    const val TOLLERANZA_FASCIA_MIN = 1L

    fun valuta(
        regole: List<Regola>,
        bonusOggiPerRegola: Map<String, Int>,
        usoMinutiEtichetta: (String) -> Long,
        usoMinutiIntervallo: (Long, Long) -> Long,
        now: Long,
        zona: ZoneId,
    ): List<Sforamento> {
        val risultati = mutableListOf<Sforamento>()
        for (regola in regole) {
            if (!regola.attiva) continue
            when (regola.tipo) {
                TipiRegola.LIMITE_TEMPO ->
                    valutaLimite(regola, bonusOggiPerRegola, usoMinutiEtichetta)
                        ?.let { risultati += it }
                TipiRegola.FASCIA_ORARIA ->
                    risultati += valutaFascia(regola, usoMinutiIntervallo, now, zona)
                // vita_reale: niente sforamento automatico, si dichiara a mano.
                else -> Unit
            }
        }
        return risultati
    }

    /**
     * Il limite efficace di oggi di una limite_tempo: `minuti_al_giorno` + i bonus
     * concessi oggi su QUELLA regola (contratto, POST /api/bonus). Null se la
     * regola non è un limite di tempo o ha parametri illeggibili.
     */
    fun limiteEfficace(regola: Regola, bonusOggiPerRegola: Map<String, Int>): Int? {
        if (regola.tipo != TipiRegola.LIMITE_TEMPO) return null
        val limite = intero(regola.parametri, "minuti_al_giorno") ?: return null
        return limite + (bonusOggiPerRegola[regola.id.toString()] ?: 0)
    }

    /**
     * (v2.4) Il `giorno` dei dettagli di uno sforamento: il giorno locale del
     * telefono in cui è successo, oppure — per le fasce che scavalcano la
     * mezzanotte — il giorno di ancoraggio dell'occorrenza (lo stesso del dedup).
     * Così uno sforamento consegnato in ritardo cade nel giorno giusto.
     */
    fun giornoDelloSforamento(sforamento: Sforamento, giornoTelefono: String): String =
        sforamento.giornoAncora ?: giornoTelefono

    /** I dettagli dell'evento `sforamento` (contratto-api.md, POST /api/eventi). */
    fun dettagliSforamento(sforamento: Sforamento, giornoTelefono: String): JsonObject =
        buildJsonObject {
            put("regola_id", sforamento.regolaId)
            sforamento.limiteEfficace?.let { put("limite_efficace", it) }
            put("minuti_oltre", sforamento.minutiOltre)
            put("giorno", giornoDelloSforamento(sforamento, giornoTelefono))
        }

    private fun valutaLimite(
        regola: Regola,
        bonusOggiPerRegola: Map<String, Int>,
        usoMinutiEtichetta: (String) -> Long,
    ): Sforamento? {
        val app = stringa(regola.parametri, "app_o_categoria") ?: return null
        val limiteEfficace = limiteEfficace(regola, bonusOggiPerRegola) ?: return null
        val uso = usoMinutiEtichetta(app)
        if (uso <= limiteEfficace) return null
        return Sforamento(
            regolaId = regola.id,
            tipo = regola.tipo,
            limiteEfficace = limiteEfficace,
            minutiOltre = (uso - limiteEfficace).toInt(),
        )
    }

    /**
     * Una fascia può avere due parti visibili oggi (la coda mattutina della
     * fascia di ieri e la testa serale di quella di oggi): sono DUE occorrenze
     * diverse, quindi si raggruppano per giorno di ancoraggio e ognuna genera al
     * più uno sforamento. Le due parti della stessa occorrenza (sera + mattina
     * dopo mezzanotte) condividono invece lo stesso ancoraggio, così non si
     * contano due volte.
     */
    private fun valutaFascia(
        regola: Regola,
        usoMinutiIntervallo: (Long, Long) -> Long,
        now: Long,
        zona: ZoneId,
    ): List<Sforamento> {
        val dalle = ora(regola.parametri, "dalle") ?: return emptyList()
        val alle = ora(regola.parametri, "alle") ?: return emptyList()
        val giorni = stringhe(regola.parametri, "giorni")
        if (giorni.isEmpty()) return emptyList()

        return intervalliProibitiOggi(dalle, alle, giorni, now, zona)
            .groupBy { it.giornoAncora }
            .mapNotNull { (giornoAncora, intervalli) ->
                val usoTot = intervalli.sumOf { usoMinutiIntervallo(it.inizio, it.fine) }
                if (usoTot < TOLLERANZA_FASCIA_MIN) return@mapNotNull null
                Sforamento(
                    regolaId = regola.id,
                    tipo = regola.tipo,
                    limiteEfficace = null,
                    minutiOltre = usoTot.toInt(),
                    giornoAncora = giornoAncora,
                )
            }
    }

    /**
     * Gli intervalli proibiti da una fascia oraria che ricadono OGGI (fuso del
     * telefono) e sono già iniziati, clippati a [inizio di oggi, adesso], ognuno
     * etichettato col giorno di ANCORAGGIO (quando la fascia parte). Gestisce le
     * fasce che scavalcano la mezzanotte (es. 23:00→07:00): la parte serale
     * appartiene al giorno che la fa partire, quella mattutina al giorno prima —
     * il controllo su `giorni` e l'ancoraggio usano il giorno di partenza.
     */
    fun intervalliProibitiOggi(
        dalle: LocalTime,
        alle: LocalTime,
        giorni: List<String>,
        now: Long,
        zona: ZoneId,
    ): List<IntervalloProibito> {
        val oggi = Instant.ofEpochMilli(now).atZone(zona).toLocalDate()
        val inizioOggi = oggi.atStartOfDay(zona).toInstant().toEpochMilli()
        val risultati = mutableListOf<IntervalloProibito>()
        // Due ancoraggi bastano: la fascia di ieri (parte mattutina di oggi) e
        // quella di oggi (parte serale). Clippando a [inizioOggi, now] non c'è
        // doppio conteggio con i giorni vicini.
        for (ancora in listOf(oggi.minusDays(1), oggi)) {
            val etichetta = GIORNI[ancora.dayOfWeek.value - 1]
            if (etichetta !in giorni) continue
            val inizioLocale = ancora.atTime(dalle)
            val fineLocale =
                if (alle.isAfter(dalle)) ancora.atTime(alle) else ancora.plusDays(1).atTime(alle)
            val inizioMs = inizioLocale.atZone(zona).toInstant().toEpochMilli()
            val fineMs = fineLocale.atZone(zona).toInstant().toEpochMilli()
            val s = maxOf(inizioMs, inizioOggi)
            val e = minOf(fineMs, now)
            if (e > s) risultati += IntervalloProibito(ancora.toString(), s, e)
        }
        return risultati
    }

    /**
     * Dove sta oggi una fascia oraria rispetto ad adesso, per dirlo a parole
     * nella schermata Oggi ("mancano 3 h alle 23:00"). Stessa geometria di
     * [intervalliProibitiOggi]: la fascia appartiene al giorno in cui parte.
     * Null se la regola non è una fascia leggibile.
     */
    fun momentoFascia(regola: Regola, now: Long, zona: ZoneId): MomentoFascia? {
        if (regola.tipo != TipiRegola.FASCIA_ORARIA) return null
        val dalle = ora(regola.parametri, "dalle") ?: return null
        val alle = ora(regola.parametri, "alle") ?: return null
        val giorni = stringhe(regola.parametri, "giorni")
        val adesso = Instant.ofEpochMilli(now)
        val oggi = adesso.atZone(zona).toLocalDate()

        var passataOggi = false
        for (ancora in listOf(oggi.minusDays(1), oggi)) {
            if (GIORNI[ancora.dayOfWeek.value - 1] !in giorni) continue
            val inizio = ancora.atTime(dalle).atZone(zona).toInstant()
            val fine = (if (alle.isAfter(dalle)) ancora.atTime(alle) else ancora.plusDays(1).atTime(alle))
                .atZone(zona).toInstant()
            when {
                !adesso.isBefore(inizio) && adesso.isBefore(fine) ->
                    return MomentoFascia.InCorso(minutiTra(adesso, fine), alle)
                ancora == oggi && adesso.isBefore(inizio) ->
                    return MomentoFascia.Prima(minutiTra(adesso, inizio), dalle)
                // Finita oggi: quella di oggi, o la coda mattutina di quella di ieri.
                !fine.isAfter(adesso) && !fine.isBefore(oggi.atStartOfDay(zona).toInstant()) ->
                    passataOggi = true
            }
        }
        return if (passataOggi) MomentoFascia.Finita else MomentoFascia.NonOggi
    }

    private fun minutiTra(da: Instant, a: Instant): Long =
        java.time.Duration.between(da, a).toMinutes().coerceAtLeast(1)

    private fun stringa(parametri: JsonObject, nome: String): String? =
        (parametri[nome] as? JsonPrimitive)?.content?.takeIf { it.isNotBlank() }

    private fun intero(parametri: JsonObject, nome: String): Int? =
        (parametri[nome] as? JsonPrimitive)?.content?.toIntOrNull()

    private fun stringhe(parametri: JsonObject, nome: String): List<String> =
        (parametri[nome] as? JsonArray)
            ?.mapNotNull { (it as? JsonPrimitive)?.content }
            ?: emptyList()

    private fun ora(parametri: JsonObject, nome: String): LocalTime? {
        val testo = stringa(parametri, nome) ?: return null
        return try {
            LocalTime.parse(testo)
        } catch (e: DateTimeParseException) {
            null
        }
    }
}

File: pc/Pactum/Nucleo/Valutatore.cs
using System.Globalization;
using System.Text.Json.Nodes;

namespace Pactum.Nucleo;

/// <summary>
/// Uno sforamento rilevato sul computer (evento <c>sforamento</c>). <see cref="LimiteEfficace"/>
/// solo per limite_tempo; <see cref="GiornoAncora"/> solo per le fasce: è il giorno in cui
/// l'occorrenza parte, e fa da chiave di deduplica.
/// </summary>
public sealed record Sforamento(long RegolaId, string Tipo, int? LimiteEfficace, int MinutiOltre, string? GiornoAncora = null);

public sealed record IntervalloProibito(string GiornoAncora, long Inizio, long Fine);

/// <summary>Una fascia oraria vista adesso, per <c>GET /locale/oggi</c>.</summary>
public sealed record StatoFascia(bool AttivaOra, string? ProssimoInizio, string Fine);

/// <summary>
/// Il valutatore locale, porta di <c>Valutatore.kt</c> del telefono: logica pura,
/// l'uso arriva da due funzioni (per chiave e per intervallo). Non blocca niente
/// e non deduplica: la deduplica la fa <see cref="RegistroSforamenti"/>.
/// </summary>
public static class Valutatore
{
    private static readonly string[] Giorni = { "lun", "mar", "mer", "gio", "ven", "sab", "dom" };

    /// <summary>Sotto il minuto d'uso in una fascia = rumore di misura, non uno sforamento.</summary>
    public const long TolleranzaFasciaMin = 1;

    public static string EtichettaGiorno(DateOnly data) => Giorni[((int)data.DayOfWeek + 6) % 7];

    public static List<Sforamento> Valuta(
        IEnumerable<Regola> regole,
        IReadOnlyDictionary<string, int> bonusOggiPerRegola,
        Func<string, long> minutiChiave,
        Func<long, long, long> minutiIntervallo,
        long adessoMs,
        TimeZoneInfo zona)
    {
        var risultati = new List<Sforamento>();
        foreach (var regola in regole)
        {
            if (!regola.Attiva) continue;
            switch (regola.Tipo)
            {
                case TipiRegola.LimiteTempo:
                    var s = ValutaLimite(regola, bonusOggiPerRegola, minutiChiave);
                    if (s != null) risultati.Add(s);
                    break;
                case TipiRegola.FasciaOraria:
                    risultati.AddRange(ValutaFascia(regola, minutiIntervallo, adessoMs, zona));
                    break;
            }
        }
        return risultati;
    }

    /// <summary>Il limite di oggi: <c>minuti_al_giorno</c> + i bonus concessi oggi su QUELLA regola.</summary>
    public static int? LimiteEfficace(Regola regola, IReadOnlyDictionary<string, int> bonusOggiPerRegola)
    {
        if (regola.Tipo != TipiRegola.LimiteTempo) return null;
        var limite = regola.Intero("minuti_al_giorno");
        if (limite == null) return null;
        return limite.Value + bonusOggiPerRegola.GetValueOrDefault(regola.Id.ToString(CultureInfo.InvariantCulture));
    }

    /// <summary>Il <c>giorno</c> dello sforamento: quello del computer, o l'ancoraggio della fascia.</summary>
    public static string GiornoDelloSforamento(Sforamento s, string giornoComputer) => s.GiornoAncora ?? giornoComputer;

    /// <summary>I dettagli dell'evento <c>sforamento</c> (contratto, POST /api/eventi, v2.4).</summary>
    public static JsonObject DettagliSforamento(Sforamento s, string giornoComputer)
    {
        var d = new JsonObject { ["regola_id"] = s.RegolaId };
        if (s.LimiteEfficace != null) d["limite_efficace"] = s.LimiteEfficace.Value;
        d["minuti_oltre"] = s.MinutiOltre;
        d["giorno"] = GiornoDelloSforamento(s, giornoComputer);
        return d;
    }

    private static Sforamento? ValutaLimite(Regola regola, IReadOnlyDictionary<string, int> bonus, Func<string, long> minutiChiave)
    {
        var chiave = regola.Stringa("app_o_categoria");
        if (chiave == null) return null;
        var limite = LimiteEfficace(regola, bonus);
        if (limite == null) return null;
        long uso = minutiChiave(chiave);
        if (uso <= limite.Value) return null;
        return new Sforamento(regola.Id, regola.Tipo, limite.Value, (int)(uso - limite.Value));
    }

    /// <summary>
    /// Oggi una fascia può avere due parti (la coda mattutina di quella di ieri e la
    /// testa serale di quella di oggi): sono due occorrenze, raggruppate per giorno di
    /// ancoraggio; ognuna dà al più uno sforamento.
    /// </summary>
    private static IEnumerable<Sforamento> ValutaFascia(Regola regola, Func<long, long, long> minutiIntervallo, long adessoMs, TimeZoneInfo zona)
    {
        var dalle = regola.Ora("dalle");
        var alle = regola.Ora("alle");
        var giorni = regola.Stringhe("giorni");
        if (dalle == null || alle == null || giorni.Count == 0) yield break;

        foreach (var gruppo in IntervalliProibitiOggi(dalle.Value, alle.Value, giorni, adessoMs, zona).GroupBy(i => i.GiornoAncora))
        {
            long uso = gruppo.Sum(i => minutiIntervallo(i.Inizio, i.Fine));
            if (uso < TolleranzaFasciaMin) continue;
            yield return new Sforamento(regola.Id, regola.Tipo, null, (int)uso, gruppo.Key);
        }
    }

    /// <summary>
    /// Gli intervalli proibiti di OGGI già cominciati, tagliati a [inizio di oggi, adesso],
    /// ciascuno col giorno di ancoraggio. La parte serale di una fascia che scavalca la
    /// mezzanotte appartiene al giorno che la fa partire, la mattutina al giorno prima.
    /// </summary>
    public static List<IntervalloProibito> IntervalliProibitiOggi(TimeOnly dalle, TimeOnly alle, IReadOnlyCollection<string> giorni, long adessoMs, TimeZoneInfo zona)
    {
        var oggi = Tempo.DataDi(adessoMs, zona);
        long inizioOggi = Tempo.InizioGiorno(oggi, zona);
        var risultati = new List<IntervalloProibito>();
        foreach (var ancora in new[] { oggi.AddDays(-1), oggi })
        {
            if (!giorni.Contains(EtichettaGiorno(ancora))) continue;
            var (inizioMs, fineMs) = Occorrenza(ancora, dalle, alle, zona);
            long s = Math.Max(inizioMs, inizioOggi);
            long e = Math.Min(fineMs, adessoMs);
            if (e > s) risultati.Add(new IntervalloProibito(Tempo.Testo(ancora), s, e));
        }
        return risultati;
    }

    /// <summary>
    /// Dove sta una fascia adesso: se è in corso, quando riparte e quando finisce.
    /// Stessa geometria di <see cref="IntervalliProibitiOggi"/>. Null se la regola non è una fascia leggibile.
    /// </summary>
    public static StatoFascia? Stato(Regola regola, long adessoMs, TimeZoneInfo zona)
    {
        if (regola.Tipo != TipiRegola.FasciaOraria) return null;
        var dalle = regola.Ora("dalle");
        var alle = regola.Ora("alle");
        if (dalle == null || alle == null) return null;
        var giorni = regola.Stringhe("giorni");
        var oggi = Tempo.DataDi(adessoMs, zona);

        bool attiva = false;
        foreach (var ancora in new[] { oggi.AddDays(-1), oggi })
        {
            if (!giorni.Contains(EtichettaGiorno(ancora))) continue;
            var (inizio, fine) = Occorrenza(ancora, dalle.Value, alle.Value, zona);
            if (adessoMs >= inizio && adessoMs < fine) attiva = true;
        }

        string? prossimo = null;
        for (int d = 0; d <= 7 && prossimo == null; d++)
        {
            var giorno = oggi.AddDays(d);
            if (!giorni.Contains(EtichettaGiorno(giorno))) continue;
            if (Occorrenza(giorno, dalle.Value, alle.Value, zona).Inizio > adessoMs) prossimo = Testo(dalle.Value);
        }
        return new StatoFascia(attiva, prossimo, Testo(alle.Value));
    }

    private static (long Inizio, long Fine) Occorrenza(DateOnly ancora, TimeOnly dalle, TimeOnly alle, TimeZoneInfo zona)
    {
        long inizio = Tempo.UtcMsDi(ancora, dalle, zona);
        long fine = alle > dalle ? Tempo.UtcMsDi(ancora, alle, zona) : Tempo.UtcMsDi(ancora.AddDays(1), alle, zona);
        return (inizio, fine);
    }

    private static string Testo(TimeOnly ora) => ora.ToString("HH:mm", CultureInfo.InvariantCulture);
}

/// <summary>
/// Massimo UNO sforamento per regola per giorno (per le fasce, per giorno di
/// ancoraggio): la memoria di quelli già segnalati. Si salva su disco da chi la usa.
/// </summary>
public sealed class RegistroSforamenti
{
    /// <summary>giorno → id delle regole già segnalate quel giorno.</summary>
    public Dictionary<string, List<long>> Segnalati { get; set; } = new();

    public bool Contiene(long regolaId, string giorno) =>
        Segnalati.TryGetValue(giorno, out var ids) && ids.Contains(regolaId);

    public void Aggiungi(long regolaId, string giorno)
    {
        if (!Segnalati.TryGetValue(giorno, out var ids))
        {
            ids = new List<long>();
            Segnalati[giorno] = ids;
        }
        if (!ids.Contains(regolaId)) ids.Add(regolaId);
    }

    /// <summary>Dimentica i giorni più vecchi di <paramref name="primoDaTenere"/>.</summary>
    public void Pota(DateOnly primoDaTenere)
    {
        foreach (var giorno in Segnalati.Keys.ToList())
        {
            if (!Tempo.ProvaGiorno(giorno, out var d) || d < primoDaTenere) Segnalati.Remove(giorno);
        }
    }

    /// <summary>
    /// Gli sforamenti mai segnalati prima, già registrati come segnalati: il chiamante
    /// li accoda e li annuncia. Stessa regola del telefono (SentinellaPatto).
    /// </summary>
    public List<Sforamento> Nuovi(IEnumerable<Sforamento> sforamenti, string giornoComputer)
    {
        var nuovi = new List<Sforamento>();
        foreach (var s in sforamenti)
        {
            var giorno = Valutatore.GiornoDelloSforamento(s, giornoComputer);
            if (Contiene(s.RegolaId, giorno)) continue;
            Aggiungi(s.RegolaId, giorno);
            nuovi.Add(s);
        }
        return nuovi;
    }
}
